src/main/java/redis/clients/jedis/JedisClusterCommand.java

```diff
@@ -100,131 +101,91 @@ public T runWithAnyNode() {
     }
   }
 
-  private boolean shouldBackOff(int attemptsLeft) {
-    int attemptsDone = maxAttempts - attemptsLeft;
-    return attemptsDone >= maxAttempts / 3;
-  }
-
-  private static long getBackoffSleepMillis(int attemptsLeft, Instant deadline) {
-    if (attemptsLeft <= 0) {
-      return 0;
-    }
-
-    long millisLeft = Duration.between(Instant.now(), deadline).toMillis();
-    if (millisLeft < 0) {
-      throw new JedisClusterMaxAttemptsException("Deadline exceeded");
-    }
-
-    return millisLeft / (attemptsLeft * attemptsLeft);
-  }
-
   private T runWithRetries(final int slot) {
     Instant deadline = Instant.now().plus(maxTotalRetriesDuration);
-    Supplier<Jedis> connectionSupplier = () -> connectionHandler.getConnectionFromSlot(slot);
-
-    // If we got one redirection, stick with that and don't try anything else
-    Supplier<Jedis> redirectionSupplier = null;
 
-    for (int currentAttempt = 0; currentAttempt < this.maxAttempts; currentAttempt++) {
+    JedisRedirectionException redirect = null;
+    AtomicInteger consecutiveConnectionFailures = new AtomicInteger();
+    for (int attemptsLeft = this.maxAttempts; attemptsLeft > 0; attemptsLeft--) {
       Jedis connection = null;
       try {
-        if (redirectionSupplier != null) {
-          connection = redirectionSupplier.get();
+        if (redirect != null) {
+          connection = connectionHandler.getConnectionFromNode(redirect.getTargetNode());
```

Comment:
What about this case?

We have one `master` and one `replica`.

1. `master` goes down permanently
2. We decide to ask `replica`
3. `replica` has not yet noticed that `master` is down, so it redirects us to `master`

As I read this code, we never `null` out the `redirect` variable, so we'll just keep trying to connect to the `master` node, which will never come back up.

Instead, I think we should `null` out `redirect` on `JedisConnectionException`s.

And add a test case for this :)

Reply:
updated here https://github.com/walles/jedis/pull/2/files#diff-58ec82b0d7b29da6455767ace17ef1c0d26d6e9e539480f1f1cb9b8de86bd516R130-R133